store/src/java/com/zimbra/cs/event/SolrEventStore.java

```diff
@@ -58,6 +71,93 @@ protected void deleteEventsByDataSource(String dataSourceId) throws ServiceExcep
+            default:
+                return getQueryToSearchContactAsSenderOrReceiver(contact).toString();
+        }
+    }
+
+    private String getContactFrequencyQueryForTimeRange(ContactAnalytics.TimeRange timeRange) {
+        if(ContactAnalytics.TimeRange.FOREVER.equals(timeRange)) {
+            return new MatchAllDocsQuery().toString();
+        }
+        TermRangeQuery rangeQuery = TermRangeQuery.newStringRange(LuceneFields.L_EVENT_TIME, timeRange.getSolrStartDate(), timeRange.getSolrEndDate(), true, true);
+        return rangeQuery.toString();
+    }
+
+    @Override
+    public List<ContactFrequencyGraphDataPoint> getContactFrequencyGraph(String contact, Timestamp startDate, Timestamp endDate, String aggregationBucket) throws ServiceException {
+        SolrQuery solrQuery = new SolrQuery();
+        solrQuery.setQuery(new MatchAllDocsQuery().toString());
+        solrQuery.addDateRangeFacet(LuceneFields.L_EVENT_TIME, startDate, endDate, aggregationBucket);
+        solrQuery.addFilterQuery(getQueryToSearchContactAsSenderOrReceiver(contact).toString());
+
+        if (solrHelper.needsAccountFilter()) {
+            solrQuery.addFilterQuery(getAccountFilter(accountId));
+        }
+
+        QueryResponse response = (QueryResponse) solrHelper.executeRequest(accountId, solrQuery);
+
+        List<ContactFrequencyGraphDataPoint> graphDataPoints = Collections.emptyList();
+        List<RangeFacet> facetRanges = response.getFacetRanges();
+        if(facetRanges != null && facetRanges.size() > 0) {
+            List<RangeFacet.Count> rangeFacetResult = facetRanges.get(0).getCounts();
+            graphDataPoints = new ArrayList<>(rangeFacetResult.size());
+            for (RangeFacet.Count rangeResult : rangeFacetResult) {
+                graphDataPoints.add(new ContactFrequencyGraphDataPoint(rangeResult.getValue(), rangeResult.getCount()));
+            }
+        }
+        return graphDataPoints;
+    }
+
+    private BooleanQuery getQueryToSearchContactAsSenderOrReceiver(String contact) {
+        BooleanQuery sentEventWithContactInReceiverField = searchContactForAnEventTypeInAContextField(contact, Event.EventType.SENT, Event.EventContextField.RECEIVER);
+        BooleanQuery receivedEventWithContactInSenderField = searchContactForAnEventTypeInAContextField(contact, Event.EventType.RECEIVED, Event.EventContextField.SENDER);
+
+        BooleanQuery.Builder searchForContactAsSenderOrReceiver = new BooleanQuery.Builder();
+        searchForContactAsSenderOrReceiver.add(sentEventWithContactInReceiverField, Occur.SHOULD);
+        searchForContactAsSenderOrReceiver.add(receivedEventWithContactInSenderField, Occur.SHOULD);
+
+        return searchForContactAsSenderOrReceiver.build();
+    }
+
+    private BooleanQuery searchContactForAnEventTypeInAContextField(String contact, Event.EventType eventType, Event.EventContextField eventContextField) {
```

Comment:
This is somewhat of a confusing method name. Perhaps change to something like `searchContactByEventType`?